locker/src/main/java/org/killbill/commons/locker/zookeeper/ZooKeeperGlobalLocker.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2017 The Billing Project, LLC
+ *
+ * The Billing Project licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.commons.locker.zookeeper;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.ZooKeeper;
+import org.killbill.commons.locker.GlobalLock;
+import org.killbill.commons.locker.GlobalLocker;
+import org.killbill.commons.locker.GlobalLockerBase;
+import org.killbill.commons.locker.ResetReentrantLockCallback;
+
+public class ZooKeeperGlobalLocker extends GlobalLockerBase implements GlobalLocker {
+
+    final CountDownLatch connectedSignal = new CountDownLatch(1);
+    protected static final String zkPath = "/kb/locks/accounts/";
+
+    public ZooKeeperGlobalLocker() {
+        super(null, null, DEFAULT_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    protected GlobalLock getGlobalLock(final Connection connection, final String lockName, final ResetReentrantLockCallback resetCb) {
+        try {
+            // TODO jgomez 1: use the same timeout value?
+            ZooKeeper zooKeeper = new ZooKeeper("connectString", Long.valueOf(timeout).intValue(), new Watcher() {
```

Comment:
Couple of remarks here:
1. I (would) think initialization of `ZooKeeper` should not be done each time we get the lock, but rather in the constructor. Now, that requires checking that 1/ such objects can safely be shared among different threads and 2/ Make sure there is code to handle cases when connection breaks, to re-initialize the object. In a first prototype implementation this could be omitted for simplicity.

2. The use the the `CountDownLatch` `connectedSignal` does not seem correct. The whole intent is to synchronize the main thread in this method with the one handling the callback. Therefore this main thread should call `await()`:

```
public ZooKeeper connect(String hosts, int sessionTimeout)
        throws IOException, InterruptedException {
  final CountDownLatch connectedSignal = new CountDownLatch(1);
  ZooKeeper zk = new ZooKeeper(hosts, sessionTimeout, new Watcher() {
    @Override
    public void process(WatchedEvent event) {
      if (event.getState() == Watcher.Event.KeeperState.SyncConnected) {
        connectedSignal.countDown();
      }
    }
  });
  connectedSignal.await();
  return zk;
}

```